File: core/src/com/mygdx/game/Entity/EnemyEntity.java
package com.mygdx.game.Entity;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.Texture;
import com.badlogic.gdx.graphics.g2d.Animation;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.math.Vector2;

/**
 * Created by Adrian on 2017-04-08.
 */

public class EnemyEntity extends BasicEntity {
    private int wall;
    private float timer;
    public float rotate;
    public boolean canMoving = true;
    private static final int FRAME_COLS = 17;
    Animation<TextureRegion> walkAnimation;
    Texture walkSheet;
    float stateTime;

    public EnemyEntity(String name) {
        super(name);
        wall = generator.nextInt(4);
        getRandomPosition();
        SPEED = 100;
        //pb = new ProgressBar(position.x, position.y);

        walkSheet = new Texture(Gdx.files.internal("zombiewalk.png"));
        TextureRegion[][] tmp = TextureRegion.split(walkSheet, (walkSheet.getWidth() / FRAME_COLS), walkSheet.getHeight());
        TextureRegion[] walkFrames = new TextureRegion[FRAME_COLS];

        int index = 0;
        for (int i = 0; i < FRAME_COLS; i++) {
            walkFrames[index] = tmp[0][i];
            index++;
        }
        walkAnimation = new Animation<>(0.033f, walkFrames);
        stateTime = 0f;
    }

    public void draw(SpriteBatch batch, float delta) {
        stateTime += delta;
        if(canMoving) {
            TextureRegion currentFrame = walkAnimation.getKeyFrame(stateTime, true);
            batch.draw(currentFrame,position.x, position.y, sprite.getOriginX(), sprite.getOriginY(), (walkSheet.getWidth() / FRAME_COLS)/3, walkSheet.getHeight()/3, 1, 1, sprite.getRotation()+90);
        } else {
            sprite.setPosition(position.x, position.y);
            sprite.draw(batch);
        }

    }

    public void getRandomPosition() {
        if(wall == 0) {
            position.x = 0;
            position.y = generator.nextInt(1000);
            sprite.setRotation(270);
        } else if (wall == 1) {
            position.x = generator.nextInt(1150);
            position.y = 0;
        } else if (wall == 2) {
            position.x = 1150;
            position.y = generator.nextInt(900);
            sprite.setRotation(90);
        } else  if (wall == 3) {
            position.x = generator.nextInt(1100);
            position.y = 1000;
            sprite.setRotation(180);
        }
    }

    public void update(float delta) {
        timer += delta;
        if(canMoving) {
            forward(delta);
            sprite.setRotation(sprite.getRotation() + rotate * delta);
        }

        if(timer > 2){
            rotate = generator.nextInt(100)-50;
            timer = 0;
        }
    }

    @Override
    public void dispose () {
        tex.dispose();
    }

    public static class ProgressBar extends Sprite {
        private Texture tex;
        Sprite progressbar;
        Vector2 positionOfBar;

        public ProgressBar(float x, float y){
            tex = new Texture("healthbar.png");
            progressbar = new Sprite(tex);
            positionOfBar = new Vector2(x, y);
            progressbar.setSize(50, 5);
        }

        public void draw(SpriteBatch batch) {
            progressbar.setPosition(positionOfBar.x, positionOfBar.y);
            progressbar.draw(batch);
        }

        public void reduceBar(int howMuch) {
            progressbar.setSize(progressbar.getWidth() - howMuch, progressbar.getHeight());
        }

        public void dispose() {
            tex.dispose();
        }
    }
}

File: core/src/com/mygdx/game/Entity/EnemyEntity.kt
package com.mygdx.game.Entity

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.Texture
import com.badlogic.gdx.graphics.g2d.Animation
import com.badlogic.gdx.graphics.g2d.Sprite
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.badlogic.gdx.math.Vector2

/**
 * Created by Adrian on 2017-04-08.
 */

class EnemyEntity(name: String) : MovableEntity(name, 0f, 0f) {
    private val wall: Int = BasicEntity.generator.nextInt(4)
    private var timer: Float = 0.toFloat()
    var rotate: Float = 0.toFloat()
    var canMoving = true
    internal var walkAnimation: Animation<TextureRegion>
    internal var walkSheet: Texture
    internal var stateTime: Float = 0.toFloat()

    init {
        getRandomPosition()
        SPEED = (BasicEntity.generator.nextInt(60) + 40).toFloat()
        //pb = new ProgressBar(position.x, position.y);

        walkSheet = Texture(Gdx.files.internal("zombiewalk.png"))
        val tmp = TextureRegion.split(walkSheet, walkSheet.width / FRAME_COLS, walkSheet.height)
        val walkFrames = arrayOfNulls<TextureRegion>(FRAME_COLS)

        var index = 0
        for (i in 0 until FRAME_COLS) {
            walkFrames[index] = tmp[0][i]
            index++
        }
        walkAnimation = Animation<TextureRegion>(0.033f, *walkFrames)
        stateTime = 0f
    }

    fun draw(batch: SpriteBatch, delta: Float) {
        stateTime += delta
        if (canMoving) {
            val currentFrame = walkAnimation.getKeyFrame(stateTime, true)
            batch.draw(currentFrame, position.x, position.y, sprite.originX, sprite.originY,
                    (walkSheet.width / FRAME_COLS / 4).toFloat(), (walkSheet.height / 4).toFloat(), 1f, 1f, sprite.rotation + 90)
        } else {
            sprite.setPosition(position.x, position.y)
            sprite.draw(batch)
        }

    }

    fun getRandomPosition() {
        if (wall == 0) {
            position.x = 0f
            position.y = BasicEntity.generator.nextInt(800).toFloat()
            sprite.rotation = 270f
        } else if (wall == 1) {
            position.x = BasicEntity.generator.nextInt(1280).toFloat()
            position.y = 0f
        } else if (wall == 2) {
            position.x = 1280f
            position.y = BasicEntity.generator.nextInt(800).toFloat()
            sprite.rotation = 90f
        } else if (wall == 3) {
            position.x = BasicEntity.generator.nextInt(1280).toFloat()
            position.y = 800f
            sprite.rotation = 180f
        }
    }

    fun update(delta: Float) {
        timer += delta
        if (canMoving) {
            forward(delta)
            sprite.rotation = sprite.rotation + rotate * delta
        }

        if (timer > 2) {
            rotate = (BasicEntity.generator.nextInt(100) - 50).toFloat()
            timer = 0f
        }
    }

    override fun dispose() {
        tex.dispose()
    }

    class ProgressBar(x: Float, y: Float) : Sprite() {
        private val tex: Texture
        internal var progressbar: Sprite
        internal var positionOfBar: Vector2

        init {
            tex = Texture("healthbar.png")
            progressbar = Sprite(tex)
            positionOfBar = Vector2(x, y)
            progressbar.setSize(50f, 5f)
        }

        fun draw(batch: SpriteBatch) {
            progressbar.setPosition(positionOfBar.x, positionOfBar.y)
            progressbar.draw(batch)
        }

        fun reduceBar(howMuch: Int) {
            progressbar.setSize(progressbar.width - howMuch, progressbar.height)
        }

        fun dispose() {
            tex.dispose()
        }
    }

    companion object {
        private val FRAME_COLS = 17
    }
}
